pkg/scheduler/reconcile_on_edgeplacement.go

```diff
@@ -21,79 +21,140 @@ import (
+			return err
 		}
+	}
+
+	// 1)
+	locsAll, err := c.locationLister.List(labels.Everything())
+	if err != nil {
+		logger.Error(err, "failed to list Locations in all workspaces")
 		return err
 	}
+	locsFilteredByEp, err := filterLocsByEp(locsAll, ep)
+	if err != nil {
+		logger.Error(err, "failed to find Locations for EdgePlacement")
+	}
+	locsSelecting := packLocKeys(locsFilteredByEp)
+
+	singles := []edgev1alpha1.SinglePlacement{}
+	for _, loc := range locsFilteredByEp {
+		// 2)
+		lws := logicalcluster.From(loc)
+		stsInLws, err := c.synctargetLister.Cluster(lws).List(labels.Everything())
+		if err != nil {
+			logger.Error(err, "failed to list SyncTargets in Location workspace", "locationWorkspace", lws.String())
+			return err
+		}
+		stsSelecting, err := filterStsByLoc(stsInLws, loc)
+		if err != nil {
+			logger.Error(err, "failed to find SyncTargets for Location", "locationWorkspace", lws.String(), "location", loc.Name)
+			return err
+		}
+		singles = append(singles, makeSinglePlacementsForLoc(loc, stsSelecting)...)
+	}
+
+	// 3)
+	for loc, eps := range store.epsBySelectedLoc {
+		if _, ok := locsSelecting[loc]; !ok {
+			delete(eps, epKey)
+		}
+	}
+	for loc := range locsSelecting {
+		if store.epsBySelectedLoc[loc] == nil {
+			store.epsBySelectedLoc[loc] = map[string]empty{epKey: {}}
+		} else {
+			store.epsBySelectedLoc[loc][epKey] = empty{}
+		}
+	}
 
-	_, err = c.edgeClusterClient.EdgeV1alpha1().SinglePlacementSlices().Cluster(ws.Path()).Get(ctx, name, metav1.GetOptions{})
+	// 4)
+	currentSPS, err := c.singlePlacementSliceLister.Cluster(epws).Get(epName)
 	if err != nil {
-		if errors.IsNotFound(err) {
+		if errors.IsNotFound(err) { // create
 			logger.V(1).Info("creating SinglePlacementSlice")
 			sps := &edgev1alpha1.SinglePlacementSlice{
 				ObjectMeta: metav1.ObjectMeta{
-					Name: name,
+					Name: epName,
 					OwnerReferences: []metav1.OwnerReference{
 						{
 							APIVersion: edgev1alpha1.SchemeGroupVersion.String(),
 							Kind:       "EdgePlacement",
-							Name:       name,
+							Name:       epName,
 							UID:        ep.UID,
 						},
 					},
 				},
-				Destinations: []edgev1alpha1.SinglePlacement{},
+				Destinations: singles,
 			}
-			_, err = c.edgeClusterClient.Cluster(ws.Path()).EdgeV1alpha1().SinglePlacementSlices().Create(ctx, sps, metav1.CreateOptions{})
+			_, err = c.edgeClusterClient.Cluster(epws.Path()).EdgeV1alpha1().SinglePlacementSlices().Create(ctx, sps, metav1.CreateOptions{})
 			if err != nil {
 				if !errors.IsAlreadyExists(err) {
-					logger.Error(err, "failed creating singlePlacementSlice")
+					logger.Error(err, "failed creating SinglePlacementSlice")
 					return err
 				}
 			} else {
 				logger.V(1).Info("created SinglePlacementSlice")
 			}
 		} else {
-			logger.Error(err, "failed getting SinglePlacementSlice for EdgePlacement")
+			logger.Error(err, "failed getting SinglePlacementSlice")
 			return err
 		}
+	} else { // update
+		currentSPS.Destinations = singles
```

Comment:
If every change notification provokes a write, and remember that every write will provoke a change notification unless there is a communication problem, won't this produce an infinite loop?